### PR TITLE
Issue #214 Delegate the CacheAccessException thrown from the cache it…

### PR DESCRIPTION
**Pull request by albinsuresh**
…erator to the resilience strategy

### CODE REVIEW

File: core/src/main/java/org/ehcache/Ehcache.java [hunk 1]
```diff
@@ -1379,9 +1383,13 @@ public boolean remove(K key) {
...
-      for (Iterator<Cache.Entry<K, V>> iter = new CacheEntryIterator(true); iter.hasNext(); ) {
-        K key = iter.next().getKey();
-        remove(key);
+      try {
+        for (Iterator<Entry<K, V>> iter = new CacheEntryIterator(true); iter.hasNext(); ) {
+          K key = iter.next().getKey();
+          remove(key);
+        }
+      } catch (CacheAccessException e) {
+        throw new RuntimeException(e);
```
Comment (albinsuresh) [inline core/src/main/java/org/ehcache/Ehcache.java]: Throwing RuntimeException to be consistent with the behaviour of the other methods. If we decide to handle these too with resilience strategy, that will be done as a separate commit.

**Comment by albinsuresh** [comment]
Transformation of RuntimeExceptions from the store level to CacheAccessExceptions will be done as a separate commit. Raising the PR with just this commit to get some feedback on the resilience strategy split approach.

**Comment by ljacomet** [comment]
Had a discussion with @albinsuresh and I recommend that the underlying Store iterators be adapted to remember when they encounter a problem and only throw on `Store.Iterator.next()` which will then simplify the resilience strategy interactions by having a single method that could return an entry if a user later chooses to do so but throw by default.

File: api/src/main/java/org/ehcache/exceptions/CachePassThroughException.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+package org.ehcache.exceptions;
+
+/**
+ * A generic wrapper runtime exception that will not be caught and
+ * handled at the store level. This wrapper can be used to leak any
+ * runtime exceptions that you don't want to be caught at the store level.
+ * 
+ * @author Albin Suresh
+ */
+public class CachePassThroughException extends RuntimeException {
```
Comment (ljacomet) [inline api/src/main/java/org/ehcache/exceptions/CachePassThroughException.java]: I do not think we want that type in the API module, it is more of an internal implementation concern. Can it live in `impl` (ideal) or `core`?

File: core-spi-test/src/main/java/org/ehcache/internal/store/StorePutIfAbsentTest.java [hunk 3]
```diff
@@ -143,9 +144,8 @@ public void wrongKeyTypeThrowsException()
         kvStore2.putIfAbsent("key", value);
       }
       throw new AssertionError("Expected ClassCastException because the key is of the wrong type");
-    } catch (ClassCastException e) {
-      // expected
-    } catch (CacheAccessException e) {
+    } catch (CacheAccessException cae) {
+      assertThat(cae.getCause(), instanceOf(ClassCastException.class));
```
Comment (ljacomet) [inline core-spi-test/src/main/java/org/ehcache/internal/store/StorePutIfAbsentTest.java]: Wait, that changes the test semantics. You should do nothing if the cause is a `ClassCastException` and throw otherwise.

File: core-spi-test/src/main/java/org/ehcache/internal/store/StorePutIfAbsentTest.java [hunk 4]
```diff
@@ -165,9 +165,8 @@ public void wrongValueTypeThrowsException()
         kvStore2.putIfAbsent(key, "value");
       }
       throw new AssertionError("Expected ClassCastException because the value is of the wrong type");
-    } catch (ClassCastException e) {
-      // expected
-    } catch (CacheAccessException e) {
+    } catch (CacheAccessException cae) {
+      assertThat(cae.getCause(), instanceOf(ClassCastException.class));
```
Comment (ljacomet) [inline core-spi-test/src/main/java/org/ehcache/internal/store/StorePutIfAbsentTest.java]: Wait, that changes the test semantics. You should do nothing if the cause is a `ClassCastException` and throw otherwise.

File: core/src/test/java/org/ehcache/EhcacheBasicGetTest.java [hunk 5]
```diff
@@ -148,7 +150,8 @@ public void testGetNoStoreEntryCacheLoadingException() throws Exception {
     try {
       ehcache.get("key");
       fail();
-    } catch (CacheLoadingException e) {
+    } catch (CachePassThroughException e) {
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetTest.java]: Ehcache should not leak `CachePassThroughException`, this should still work by catching the `CacheLoadingException`

Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetTest.java]: Looking a bit more at the code, that probably means doing some work on the `FakeStore` used here to match what happened in the real stores.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 6]
```diff
@@ -698,64 +831,86 @@ public String toString() {
...
-        if (mappedValue != null) {
-          long expirationTime = mappedValue.expirationTime(OnHeapValueHolder.TIME_UNIT);
-          return newUpdateValueHolder(key, existingValue, computedValue, now, expirationTime);
-        } else {
-          return newCreateValueHolder(key, computedValue, now);
         }
+      });
+      return enforceCapacityIfValueNotNull(computeResult);
+    } catch (CachePassThroughException cpte) {
+      Throwable cause = cpte.getCause();
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: These few lines are repeated all over the place and in the `AbstractOffHeapStore` as well.
So they should be extracted. What is also interesting is that none of this requires the context of `this` class, so it would be a valid candidate for a method on `CachePassThroughException`.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 7]
```diff
@@ -401,26 +491,34 @@ private void invalidate() {
...
-          next = entry;
+        } catch (RuntimeException re) {
+          prefetchFailure = new CacheAccessException(re);
         }
       }
       
       @Override
-      public boolean hasNext() throws CacheAccessException {
+      public boolean hasNext() {
         return next != null;
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: I believe this should become `return next != null || prefetchFailure != null`

Comment (albinsuresh) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: But that is wrong, no? _hasNext_ is supposed to return whether or not a next element is there. A prefetchFailure happens when a next element "is there" but we just failed to retrieve that. So _hasNext_ returning false in that case is like lying to the user, right? The next element is actually there. Even though there is no value in returning *true* either since that is not useful to the user in any way.

Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: According to my reading of the code, if you have a `prefetchFailure`, then `next` is null. So in order to have `hasNext()` return true, it should check both `next` and `prefetchFailure` and either different than `null` indicates there is/was something ... a real one or an exception waiting to happen.

File: impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java [hunk 8]
```diff
@@ -842,25 +1013,35 @@ public boolean test(Map.Entry<K, OffHeapValueHolder<V>> argument) {
...
+        prefetchFailure = ce;
+      } catch (RuntimeException re) {
+        prefetchFailure = new CacheAccessException(re);
       }
     }
 
     @Override
-    public boolean hasNext() throws CacheAccessException {
+    public boolean hasNext() {
       return next != null;
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java]: See `OnHeapStore` comment

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java (hunk 7 above)
Comment (albinsuresh) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: Aah yes. My bad, It is an OR of both the conditions. I misread that as an AND.

**Comment by albinsuresh** [comment]
Committed the exception handling code refactoring changes.
